velox/exec/SortBuffer.cpp

```diff
@@ -0,0 +1,360 @@
+  if (currentDataRows == 0) {
+    // 'data_' is empty. Nothing to spill.
+    return;
+  }
+
+  auto [freeRows, outOfLineFreeBytes] = data_->freeSpace();
+  const auto outOfLineBytes =
+      data_->stringAllocator().retainedSize() - outOfLineFreeBytes;
+  const int64_t outOfLineBytesPerRow = outOfLineBytes / currentDataRows;
+  const int64_t flatInputBytes = input->estimateFlatSize();
+
+  // Test-only spill path.
+  if (currentDataRows > 0 && spillConfig_->testSpillPct &&
+      (folly::hasher<uint64_t>()(++spillTestCounter_)) % 100 <=
+          spillConfig_->testSpillPct) {
+    const int64_t rowsToSpill = std::max<int64_t>(1, currentDataRows / 10);
+    spill(
+        currentDataRows - rowsToSpill,
+        outOfLineBytes - (rowsToSpill * outOfLineBytesPerRow));
+    return;
+  }
+
+  // If current memory usage exceeds spilling threshold, trigger spilling.
+  const auto currentMemoryUsage = pool_->currentBytes();
+  if (spillMemoryThreshold_ != 0 &&
+      currentMemoryUsage > spillMemoryThreshold_) {
+    const int64_t bytesToSpill =
+        currentMemoryUsage * spillConfig_->spillableReservationGrowthPct / 100;
+    auto rowsToSpill = std::max<int64_t>(
+        1, bytesToSpill / (data_->fixedRowSize() + outOfLineBytesPerRow));
+    spill(
+        std::max<int64_t>(0, currentDataRows - rowsToSpill),
+        std::max<int64_t>(
+            0, outOfLineBytes - (rowsToSpill * outOfLineBytesPerRow)));
+    return;
+  }
+
+  // If we have enough free rows for input rows and enough variable length
+  // free space for the vector's flat size, no need for spilling.
+  if (freeRows > input->size() &&
+      (outOfLineBytes == 0 || outOfLineFreeBytes >= flatInputBytes)) {
+    return;
+  }
+
+  // For variable length data, we take the flat size of the input as the cap.
+  const int64_t estimatedIncrementalBytes =
+      data_->sizeIncrement(input->size(), outOfLineBytes ? flatInputBytes : 0);
+
+  // If the current available reservation in memory pool is 2X the
+  // estimatedIncrementalBytes, no need to spill.
+  if (pool_->availableReservation() > 2 * estimatedIncrementalBytes) {
+    return;
+  }
+
+  // Try reserving targetIncrementBytes more in memory pool, if succeed, no
+  // need to spill.
+  const auto targetIncrementBytes = std::max<int64_t>(
+      estimatedIncrementalBytes * 2,
+      currentMemoryUsage * spillConfig_->spillableReservationGrowthPct / 100);
+  if (pool_->maybeReserve(targetIncrementBytes)) {
+    return;
+  }
+
+  // NOTE: disk spilling use the system disk spilling memory pool instead of
+  // the operator memory pool.
+  const int64_t rowsToSpill = std::max<int64_t>(
+      1, targetIncrementBytes / (data_->fixedRowSize() + outOfLineBytesPerRow));
+  spill(
+      std::max<int64_t>(0, currentDataRows - rowsToSpill),
+      std::max<int64_t>(
+          0, outOfLineBytes - (rowsToSpill * outOfLineBytesPerRow)));
+}
+
+void SortBuffer::spill(int64_t targetRows, int64_t targetBytes) {
+  VELOX_CHECK_GE(targetRows, 0);
```

Comment:
As you mentioned offline, we could add spillConfig_ != nullptr check here:
```
VELOX_CHECK_NOTNULL(spillConfig_);
if (spiller_ == nullptr) {
  ...
}
...
```

Reply:
I gave it a second thought and think I need to confirm with you, what is the expected behavior for reclaim function if spilling is not enabled
- throw exception? or
- do nothing?